src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+        if (!isDirty()) {
+            isDirty = true;
+
+            // Mark the observers as dirty as they inherit it from this instance.
+            // As it is impossible to add observers after construction then a livelock would never happen.
+            for (DeferredProperty<?> property : observers) {
+                property.markAsDirty();
+            }
+
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Sets the dirty flag as <code>false</code>. If this instance was dirty then true si removed to indicate the dirty flag change from <code>true</code> to <code>false</code>.
+     * Unlike {@link DeferredProperty#markAsDirty()} this does not propagate the cleaning to its observers. For this reason, observers that are dirty will not be cleaned.
+     *
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean() {
+        if (isDirty()) {
+            isDirty = false;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If this instance {@link DeferredProperty#isDirty} then we run the {@link Runnable} and then {@link DeferredProperty#clean()}
+     * This should be useed if the dirty state of this instance can be controled by performing an operation rather than setting a new {@link DeferredProperty#value}.
+     *
+     * @param runnable The runnable which is run if we cleaned the object.
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean(Runnable runnable) {
+        boolean cleaned = clean();
+        if (cleaned) {
+            runnable.run();
+        }
+        return cleaned;
+    }
+
+    /**
+     * Used to check if this instance is marked as dirty.
+     *
+     * @return True if this instance is dirty.
+     */
+    public boolean isDirty() {
+        return isDirty;
+    }
+
+    /**
+     * Used to check if this instance is _not_ marked as dirty.
+     *
+     * @return True if this instance is not dirty.
+     */
+    public boolean isClean() {
+        return !isDirty();
+    }
+
+    /**
+     * Set the {@link DeferredProperty#value} and cleans this instance by not marking it as dirty.
+     * Calling this function when this instance is not dirty will just set {@link DeferredProperty#value}.
+     *
+     * @param value The value to set this property {@link DeferredProperty#value} to.
+     * @return Returns <code>true</code> if the instance was cleaned by marking it as no longer dirty.
+     */
+    public boolean set(T value) {
+        this.value = value;
+        return clean();
+    }
+
+    /**
+     * Sets the {@link DeferredProperty#value} but only if this is dirty.
+     * Calling this function when this instance is not dirty will not set the value.
+     * However, the current value of this property will always be returned.
```

Comment:
Is current value the one before or after setting the value? Clarify this also in the `@return` statement.